zipline/testing/core.py

```diff
@@ -1406,3 +1168,177 @@ def make_test_handler(testcase, *args, **kwargs):
+            if old_value is None:
+                # Value was not present when we entered, so del it out if it's
+                # still present.
+                try:
+                    del os.environ[key]
+                except KeyError:
+                    pass
+            else:
+                # Restore the old value.
+                os.environ[old_key] = old_value
+
+
+class tmp_dir(object):
+    """A reentrant temporary directory context manager.
+
+    Parameters
+    ----------
+    path : str, optional
+        The path to the directory. If not given, this will be a unique name.
+
+    See Also
+    --------
+    testfixtures.TempDirectory
+    """
+    def __init__(self, path=None):
+        self.path = path
+        self._depth = 0
+        self._tmpdir = None
+
+    def __enter__(self):
+        if not self._depth:
+            assert self._tmpdir is None, 'tmpdir was not reset to None'
+            self._tmpdir = TempDirectory(path=self.path)
+        self._depth += 1
+        return self._tmpdir
+
+    def __exit__(self, *excinfo):
+        self._depth -= 1
+        if not self._depth:
+            self._tmpdir.cleanup()
+            self._tmpdir = None
+
+
+class _TmpBarReader(with_metaclass(ABCMeta, tmp_dir)):
+    """A helper for tmp_bcolz_minute_bar_reader and tmp_bcolz_daily_bar_reader.
+
+    Parameters
+    ----------
+    env : TradingEnvironment
+        The trading env.
+    days : pd.DatetimeIndex
+        The days to write for.
+    data : dict[int -> pd.DataFrame]
+        The data to write.
+    path : str, optional
+        The path to the directory to write the data into. If not given, this
+        will be a unique name.
+    """
+    @abstractproperty
+    def _reader(self):
+        raise NotImplementedError('_reader')
+
+    @abstractmethod
+    def _write(self, env, days, path, data):
+        raise NotImplementedError('_write')
+
+    def __init__(self, env, days, data, path=None):
+        super(_TmpBarReader, self).__init__(path=path)
+        self._env = env
+        self._days = days
+        self._data = data
+
+    def __enter__(self):
+        tmpdir = super(_TmpBarReader, self).__enter__()
+        env = self._env
+        self._write(
+            env,
+            self._days,
+            tmpdir.path,
+            self._data,
+        )
+        return self._reader(tmpdir.path)
+
+
+class tmp_bcolz_minute_bar_reader(_TmpBarReader):
+    """A reentrant temporary BcolzMinuteBarReader object.
+
+    Parameters
+    ----------
+    env : TradingEnvironment
+        The trading env.
+    days : pd.DatetimeIndex
+        The days to write for.
+    data : dict[int -> pd.DataFrame]
+        The data to write.
+    path : str, optional
+        The path to the directory to write the data into. If not given, this
+        will be a unique name.
+
+    See Also
+    --------
+    tmp_bcolz_daily_bar_reader
+    """
+    _reader = BcolzMinuteBarReader
```

Comment:
I might call this `_reader_cls` to be clear that it's the type, not the reader itself.